### PR TITLE
More work on `zstd` compression

### PR DESCRIPTION
**Pull request by lqd**
r? @Kobzol as we've discussed this.

This is a draft to show the current approach of supporting zstd in compiletest, and making the tests using it unconditional. 

Knowing whether llvm/lld was built with `LLVM_ENABLE_ZSTD` is quite hard, so there are two strategies. There are details in the code, and we can discuss this approach. Until we know the config used to build CI artifacts, it seems our options are somewhat limited in any case.

zlib compression seems always enabled, so we only check this in its dedicated test, allowing the test to ignore errors due to zstd not being supported. 

The zstd test is made unconditional in what it tests, by relying on `needs-llvm-zstd` to be ignored when `llvm.libzstd` isn't enabled in `config.toml`.

try-job: x86_64-gnu
try-job: x86_64-msvc
try-job: x86_64-gnu-distcheck

### CODE REVIEW

**Comment by lqd** [comment]
My understanding is that test builders use `download-ci-llvm`, which has `llvm.libzstd` enabled by default on x64 linux. 

The `x86_64-gnu` builder already has `libzstd-dev` installed but it probably shouldn't be needed if it indeed uses CI artifacts (we can check that later). 

That builder should now have `needs-llvm-libzstd` tests enabled even without enabling the config. Let's check that.

@bors try

**Comment by lqd** [comment]
(Let me know what tests and try builds we'd want to double-check.)

**Comment by lqd** [comment]
[zlib](https://github.com/rust-lang-ci/rust/actions/runs/10333315674/job/28605514096#step:25:48082)
> test [run-make] tests/run-make/compressed-debuginfo ... ok

[zstd](https://github.com/rust-lang-ci/rust/actions/runs/10333315674/job/28605514096#step:25:48085)
> test [run-make] tests/run-make/compressed-debuginfo-zstd ... ok

--- 

I've removed the dependency from the test runner image, let's see what happens now.
@bors try

**Comment by lqd** [comment]
[Cool](https://github.com/rust-lang-ci/rust/actions/runs/10334928775/job/28608909805#step:25:48082)
> test [run-make] tests/run-make/compressed-debuginfo ... ok
> test [run-make] tests/run-make/compressed-debuginfo-zstd ... ok

File: tests/run-make/compressed-debuginfo-zstd/rmake.rs [hunk 1]
```diff
@@ -0,0 +1,42 @@
+// Checks debuginfo compression both for the always-enabled zlib, and when the optional zstd is
+// enabled:
+// - via rustc's `debuginfo-compression`,
+// - and via rust-lld's `compress-debug-sections`
+
+//@ needs-llvm-zstd: we want LLVM/LLD to be built with zstd support
```
Comment (Kobzol) [inline tests/run-make/compressed-debuginfo-zstd/rmake.rs:6]: The ignore reason has an additional space, but that's a detail:
```
ignored, ignored if LLVM wasn't build with zstd for ELF section compression ( we want LLVM/LLD to be built with zstd support)
```

**Review by Kobzol — COMMENTED** [review]
This looks great! However, I'm not sure if we can ignore the LLVM build in test runners everytime. The runner uses `llvm.download-ci-llvm=if-unchanged`, so when there's a change to LLVM (~~or `rustc_llvm`~~), it will actually rebuild LLVM. In that case, we can either ignore the test (but it's not ideal, as the test would only run in the next merge), or make sure that we actually build LLVM with ZSTD support even in the test runner.

Could you try to make a fake LLVM change to see what happens?

File: tests/run-make/compressed-debuginfo-zstd/rmake.rs (hunk 1 above)
Comment (lqd) [reply tests/run-make/compressed-debuginfo-zstd/rmake.rs:6]: Ugh! I will look into where this concat appears. I cannot avoid spaces after a colon 😱 

Could you remind me where we store the list of all ignore reasons so I can quickly spot if just stripping would not regress any other ignore reasons?

Comment (Kobzol) [reply tests/run-make/compressed-debuginfo-zstd/rmake.rs:6]: Umm, I have no idea :sweat_smile: We have some test with a list of all ignore reasons?

**Comment by lqd** [comment]
> Could you try to make a fake LLVM change to see what happens?

I pushed changes to `rustc_llvm` as it's easier than messing with the llvm submodule.

Let's see what happens: @bors try

File: tests/run-make/compressed-debuginfo-zstd/rmake.rs (hunk 1 above)
Comment (lqd) [reply tests/run-make/compressed-debuginfo-zstd/rmake.rs:6]: Aren't they colocated with the list of ignored tests? And that similar list was used by pietro for the ferrocene reports which contained the ignore reasons?

**Comment by Kobzol** [comment]
Looks like it downloaded LLVM anyway :laughing: `rustc_llvm` isn't really cached on CI I think, so probably you'll actually need to modify LLVM.

File: tests/run-make/compressed-debuginfo-zstd/rmake.rs (hunk 1 above)
Comment (Kobzol) [reply tests/run-make/compressed-debuginfo-zstd/rmake.rs:6]: I created a list of currently ignored tests by running some Python scripts on our CI logs, I'm not aware of a location that actually stores them, tbh. I can try to generate that list, if you want, although I'm not sure if we also store the ignore reasons on CI (but I think that we do).

Comment (lqd) [reply tests/run-make/compressed-debuginfo-zstd/rmake.rs:6]: don't worry about it :) I'll strip whitespace and we can move on with our lives

**Comment by lqd** [comment]
Alright I pushed a different commit to the llvm fork, hopefully the builder will pick it now ....

@bors try

File: tests/run-make/compressed-debuginfo-zstd/rmake.rs (hunk 1 above)
Comment (lqd) [reply tests/run-make/compressed-debuginfo-zstd/rmake.rs:6]: Done

**Comment by Kobzol** [comment]
CI was green, but the test was skipped (as expected) :/ This is a bit annoying, we'd need to keep the dist and test builders in sync.. But probably it's the most straightforward way for now (enable `llvm.libzstd` and install the `zstd` package on the test runner).

**Comment by lqd** [comment]
Sure we can install the package on the test builder and enable zstd, but still manually build a specific version on the dist builder? Or should we build-ztsd.sh on both?

**Comment by Kobzol** [comment]
Let's install the package on the test builder, we'll at least get more coverage out of that (test that it also works with a repository-provided zstd).

**Comment by lqd** [comment]
This should be the last try to see the tests pass 🤞 

@bors try

**Comment by lqd** [comment]
[Both passed.](https://github.com/rust-lang-ci/rust/actions/runs/10342121462/job/28624697252#step:25:55338)

>   test [run-make] tests/run-make/compressed-debuginfo ... ok
  test [run-make] tests/run-make/compressed-debuginfo-zstd ... ok

**Comment by Kobzol** [comment]
Awesome! Thanks, feel free to r=me.

**Comment by lqd** [comment]
alright let's see what bors has to say

@bors r=Kobzol

File: tests/run-make/compressed-debuginfo-zstd/rmake.rs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+    // check compressed debug sections via rust-lld flag
+    prepare_and_check(to_find, |rustc| {
+        rustc.link_arg(&format!("-Wl,--compress-debug-sections={compression}"))
+    });
+}
+
+fn prepare_and_check<F: FnOnce(&mut Rustc) -> &mut Rustc>(to_find: &str, prepare_rustc: F) {
+    run_in_tmpdir(|| {
+        let mut rustc = Rustc::new();
```
Comment (Oneirical) [inline tests/run-make/compressed-debuginfo-zstd/rmake.rs:27]: nit: this can be `rustc()` instead of `Rustc::new()`.

**Comment by lqd** [comment]
Ah I was wondering what happened to this PR 😅. I rebased and fixed the silent conflicts on the tests.

@bors r=Kobzol

**Comment by GuillaumeGomez** [comment]
This PR failed in https://github.com/rust-lang/rust/pull/129126.

@bors r-

**Comment by lqd** [comment]
> (enable `llvm.libzstd` and install the `zstd` package on the test runner).

lmao @Kobzol: `setting libzstd is incompatible with download-ci-llvm` 😂  https://github.com/rust-lang-ci/rust/actions/runs/10407225287/job/28822055279#step:25:411

**Comment by Kobzol** [comment]
Ah, crap. Ideally, we need something like https://github.com/rust-lang/rust/pull/129052, but for `download-ci-llvm`.

**Comment by onur-ozkan** [comment]
> Ah, crap. Ideally, we need something like #129052, but for `download-ci-llvm`.

Good idea, filed an [issue](https://github.com/rust-lang/rust/issues/129153) for that.

File: src/tools/compiletest/src/header.rs [hunk 3]
```diff
@@ -1203,6 +1203,107 @@ pub fn extract_llvm_version_from_binary(binary_path: &str) -> Option<u32> {
...
+        let llvm_config_path = llvm_bin_dir.join("llvm-config");
+        let output = Command::new(llvm_config_path).arg("--system-libs").output().ok()?;
+        assert!(output.status.success(), "running llvm-config --system-libs failed");
+
+        let libs = String::from_utf8(output.stdout).ok()?;
+        for lib in libs.split_whitespace() {
+            if lib.ends_with("libzstd.a") && Path::new(lib).exists() {
+                return Some(());
+            }
+        }
```
Comment (jieyouxu) [inline src/tools/compiletest/src/header.rs:1227]: Question: (hopefully not) but can these libs contain whitespace in the paths themselves? Say if I have idk `C:\Users\Ferris OwO\idk_why_llvm_libs_are_here\`, would this split weirdly?

Comment (lqd) [reply src/tools/compiletest/src/header.rs:1227]: It's not the llvm libs but the system libs, but does https://github.com/rust-lang/rust/blob/a971212545766fdfe0dd68e5d968133f79944a19/compiler/rustc_llvm/build.rs#L261-L294 answer your question?

Comment (jieyouxu) [reply src/tools/compiletest/src/header.rs:1227]: Ah right, that looks fine to me, thanks!

**Comment by lqd** [comment]
Here as well, the tests seem to [have passed](https://github.com/rust-lang-ci/rust/actions/runs/10473249505/job/29004719222#step:25:49029)

> test [run-make] tests/run-make/compressed-debuginfo ... ok
   test [run-make] tests/run-make/compressed-debuginfo-zstd ... ok

So I guess this is @rustbot ready

**Comment by Kobzol** [comment]
@onur-ozkan Just wanted to check if you are fine with https://github.com/rust-lang/rust/pull/128935/commits/1649e7128752690d10e738e3f1437dd3af4d4ca0? As a sort of a workaround until we implement `builder-config` for CI LLVM. We have recently added a new LLVM ZSTD build option, but it wasn't properly tested, this PR adds proper tests, but we need to have a way to use ZSTD on the test builder even if download CI LLVM is used.

**Comment by onur-ozkan** [comment]
> @onur-ozkan Just wanted to check if you are fine with [1649e71](https://github.com/rust-lang/rust/commit/1649e7128752690d10e738e3f1437dd3af4d4ca0)? As a sort of a workaround until we implement `builder-config` for CI LLVM. We have recently added a new LLVM ZSTD build option, but it wasn't properly tested, this PR adds proper tests, but we need to have a way to use ZSTD on the test builder even if download CI LLVM is used.

It seems okay, thanks for the ping.

**Comment by Kobzol** [comment]
Ok, thanks. In that case:

@bors r+

**Comment by tgross35** [comment]
@bors r-

The `#[cfg(not(unix))]` `is_lld_built_with_zstd` arg needs a leading underscore https://github.com/rust-lang/rust/pull/129428#issuecomment-2306051695

**Comment by Kobzol** [comment]
@bors try

Just to make sure (I enabled also the previously failing windows job for the try build).

**Comment by lqd** [comment]
stage0 compiletest is already built on x86_64-msvc by now so we should be good

**Comment by lqd** [comment]
Indeed we were,
@bors r=Kobzol

**Comment by tgross35** [comment]
@bors r-

The distcheck compressed-debuginfo test failed at https://github.com/rust-lang/rust/pull/129547#issuecomment-2308800525

**Comment by lqd** [comment]
distcheck doesn’t even have zstd enabled…

**Pull request merged**